Add genre

nperrier opened the pull request:
Another big one.

EDIT: I probably should have squashed that last commit, sorry!

nperrier committed:
Add Genre model (not quite sure how I missed this file :P)

nperrier committed:
Added Genre to Track.

Also replaced lengthy Track constructor with a Builder to make it clear
what the params are when creating a Track object.

rodnaph commented on src/com/pugh/sockso/db/HSQLDatabase.java:
```diff
@@ -619,6 +620,17 @@ private void checkDefaultStructure() {
             );
             log.debug( "Created 'tracks' table" );
             
+            // genre
```
The creation of the genre tables will need to go in the _checkGenreColumnExists()_ method, suggest renaming to _checkGenreSchemaExists()_

nperrier replied:
Whoops, I forgot to rename that one.  Good catch.

rodnaph commented on src/com/pugh/sockso/music/CollectionManagerListener.java:
```diff
@@ -16,12 +16,13 @@
     public static final int ARTIST_ADDED = 1;
     public static final int ALBUM_ADDED = 2;
     public static final int TRACK_ADDED = 3;
-    public static final int PLAYLISTS_CHANGED = 4;
```
Probably best to keep existing ID's as they are (just incase) and add the genre one to the end instead.

nperrier replied:
I don't think the numbers have any meaning, but I can move it.

I'll investigate to make sure.

rodnaph commented on src/com/pugh/sockso/music/DBCollectionManager.java:
```diff
@@ -305,23 +311,25 @@ protected void checkAlbumTagInfo( final int artistId, final Tag tag, final Track
 
         // need to ignore case because that's how the DB does it
-        if ( !track.getArtist().getName().toLowerCase().equals(tag.getArtist().toLowerCase()) ) {
+        if ( !track.getArtist().getName().equalsIgnoreCase(tag.getArtist()) ) {
```
Well I never...  can't believe I didn't know this method existed, haha :)

nperrier replied:
Yeah, it was a bit more complicated than it needed to be :)

rodnaph commented on src/com/pugh/sockso/music/DBCollectionManager.java:
```diff
@@ -802,13 +862,76 @@ private int addAlbum( final int artistId, String name, String year ) {
+                Utils.close( st );
+            }
+
+            st = db.prepare(
```
Surely this needs to be in the same try/catch as the insert above?  ie. If the insert fails then we can't fetch the new ID.

We should probably _select max(id)_ as well to make sure we've got the latest ID (different DB impls might have different default orderings).

I would have assumed the _Database_ interface would have had a _lastInsertId()_ method, but it doesn't... :(  Maybe worth adding it?

nperrier replied:
I'll investigate.

nperrier replied:
I've removed the nested try/catch from the method addGenre().

I think what we really need (and this would be a big change) is transaction management.  The problem is difficult, though, because we support so many different databases.  It would be easier if we were using an ORM tool like Hibernate, but I digress.

I don't think "select max(id)" would be good because of my previous comment; we can't guarantee the query for last insert id was for the db insert just completed in a multi-threaded environment

nperrier replied:
Actually, I take that back.  After I tested it with just the outer try/catch, the return value of the method was always -1, because the insert only works once, then fails and throws an exception which skips the select id in the second statement.

rodnaph replied:
JDBC provides a transaction API using connection.setAutoCommit() - not sure it's supported by all database types, but if it's not then I don't think we should worry about it.

As for getting the last insert id, using max() is far from optimal yes i agree.  we could add something to the database interface to handle this more neatly, wrapping this possibly...

http://docs.oracle.com/javase/6/docs/api/java/sql/Statement.html#executeUpdate%28java.lang.String,%20int%29

(using the generated keys constant i mean).

rodnaph commented on src/com/pugh/sockso/music/MusicSearch.java:
```diff
@@ -83,17 +83,20 @@ protected String createQuery( final String query ) {
                     " on ar.id = t.artist_id " +
                     " inner join albums al " +
                     " on al.id = t.album_id " +
+                    " inner join genres g " +
```
Will all existing music have genres?  If a track might not have a genre we probably need to make this a _left join_ otherwise we'll lose the results from the search.

nperrier replied:
You may be right, i'll test it out and see what the results are using inner vs. outer joins

nperrier replied:
1. For new tracks, the file scanner will set `genre_id` for `track` rows with no genre as the placeholder "Unknown Genre", so all new tracks should have a genre
2. For existing tracks, a forced re-scan is needed in order to add the genre (if it exists) and set the placeholder if it doesn't

So, yes it is an issue as the `genre_id` can be null, but after a full scan there should theoretically be no nulls

rodnaph replied:
I'd rather not require a rescan (I wouldn't want someone to upgrade and then suddenly all their music dissapears! hehe).

We could do this instead by populating the default during the database update instead?  Removing the need for a rescan.  What do you think?

nperrier replied:
I think makes a bit more sense, but be warned: this would required updating EVERY track row in the database with a default "Unknown Genre".

Assuming the file scanner runs post-upgrade, the genre for a track tag (that has one) will then change from the default to the real genre for all tracks.

rodnaph replied:
Aye, I think that's fine.  I'd rather the user upgrade and see there's a new genre field that defaults to no genre, rather than upgrading and seeing all their music disappear until then run another collection scan.

nperrier replied:
Alright, I got the upgrade working pretty good.  One issue though:  After I add the genre table, update tracks to add the genre_id, and set the default "Unknown" genre on all the tracks, when I force a re-scan (through the UI), how do I force the indexer to scan files that haven't had their last-modified date changed?  My thought was to update the indexer rows and set every track to 0, but that seems like a hack.

If the indexer doesn't re-scan files that haven't changed, then their genres never update in the database to the tag values (in other words, the tracks just keep the "Unknown Genre" even thought the tag says something like "Jazz" or "Rock").

Thoughts?

nperrier replied:
{Nudge}

rodnaph replied:
Sorry for the delay. I think either...

1) add a way of customizing the indexer to specify the function to use to check if a file should be scanned.
2) use your suggestion of just resetting the last scan time on each track

First suggestion is probably quite a bit of work, and would be best as a PR in itself.  And 2 is easier...  your call, I'm ok with either as long as it's well tested.

nperrier replied:
Yes, my thought was we need a way to "force" a re-scan on files that have not changed on disk, but we want to re-read their tags for changes (since this is a special case of considering new tags that we hadn't previously (genre))

I like number 1 better, as it would definitely be useful for future changes (might as well do the real work now).

rodnaph replied:
Great!  I've created an issue with some thoughts about this in...

https://github.com/rodnaph/sockso/issues/122

Add yours and let's get this in.  I'm not happy with the name of the interface, so please think up something better! :)

nperrier replied:
Alright, now that #122 is in, any more concerns regarding this PR?

rodnaph commented on src/com/pugh/sockso/music/Track.java:
```diff
@@ -33,27 +34,84 @@
     }
 
+
+    public static class Builder {
```
I like this builder class, much neater.  Can we move it to its own file though?  (helps to keep things less coupled I find)

nperrier replied:
I could move it into a TrackBuilder class, but it seems to be tightly coupled to the Track class, which is why I made it an inner static class.

I'm following how Android devs do it: http://developer.android.com/reference/android/app/Notification.Builder.html
and how we do it at my job: http://www.crashplan.com/

I don't really have any strong feelings to where the class should belong.

rodnaph replied:
Ok, no worries.  I'll go with your call then :)

nperrier committed:
Ensure a default genre exists when server upgrades

nperrier committed:
Removed query() method from database classes

query() was returning a ResultSet from an executed Statement, but was not
explicitly closing the Statement (it was going out of scope) and this
causes issue with SQLite and MySQL.

It's good practice to explicitly close Statements when finished with them
(and ResultSet's too, though closing Statements also closes the ResultSet)

Also fixed a few issues with adding the genre entity to the data model and
a few broken tests

nperrier committed:
Add api endpoint for Genres.  Fix bug with SQL in Genre

nperrier committed:
Merge branch 'master' into add-genre

Conflicts:
	src/com/pugh/sockso/music/DBCollectionManager.java

nperrier commented:
Are you still interested in maintaining the Sockso repository?  I'd be happy to take it over for you!

Any chance we can review pull requests?

rodnaph commented:
Hey - yeah sorry but I just don't have the time now.  I should have offered it to you before really, apologies.

I've added you as a collaborator to the project, which I think will give you full push/pull access, just let me know if there's anything else I need to do.

So feel free to commit what you like, just drop me a line if you've got any questions or just want to use someone to bounce ideas off.  If you'd be interested in maintaining, or updating the website then just let me know that too.  I bought sockso.com a month or so ago just because I saw it had finally become available, so I can point that to something you set up if you like.

nperrier commented:
Cool thanks!  I've noticed you've been doing a lot of Clojure stuff lately. There's a lot of neat functional languages and JVM languages out there right now, I wish I had the time to learn the all.

Where is your current site hosted (sockso.pu-gh.com)?  You could just point the new domain to that one for now.

I noticed the website link to this github project is broken so I fixed the link.  Do you want to give me push access to the sockso site too?  

Also, the website forum isn't up it seems.  Is that no longer used?

nperrier committed:
Merge branch 'master' into add-genre

nperrier committed:
Merge branch 'master' into add-genre

nperrier committed:
Fix incorrect method call
